client/consensus/grandpa/justification.go

```diff
@@ -0,0 +1,289 @@
+			currentHash = parentHash
+		}
+	}
+
+	return Justification[H, N, S, ID, Header]{
+		Round:           round,
+		Commit:          commit,
+		VotesAncestries: voteAncestries,
+	}, nil
+}
+
+// Decode a GRANDPA justification and validate the commit and the votes'
+// ancestry proofs finalize the given block.
+func decodeAndVerifyFinalizes[H HashI, N constraints.Unsigned, S comparable, ID constraints.Ordered, Header HeaderI[H, N]](encoded []byte, finalizedTarget hashNumber[H, N], setID uint64, voters finalityGrandpa.VoterSet[ID]) (Justification[H, N, S, ID, Header], error) {
+	justification := Justification[H, N, S, ID, Header]{
+		VotesAncestries: make([]Header, 0),
+	}
+	err := scale.Unmarshal(encoded, &justification)
+	if err != nil {
+		return Justification[H, N, S, ID, Header]{}, fmt.Errorf("error decoding justification for header: %s", err)
+	}
+
+	decodedTarget := hashNumber[H, N]{
+		hash:   justification.Commit.TargetHash,
+		number: justification.Commit.TargetNumber,
+	}
+
+	if decodedTarget != finalizedTarget {
+		return Justification[H, N, S, ID, Header]{}, fmt.Errorf("invalid commit target in grandpa justification")
+	}
+
+	return justification, verifyWithVoterSet(justification, setID, voters)
+}
+
+// TODO get feedback on if I can avoid this, see below
+//type NewAuthority[ID constraints.Ordered] struct {
+//	Key    ID
+//	Weight uint64
+//}
+
+// Validate the commit and the votes' ancestry proofs.
+// func (j *Justification[H, N, S, ID]) verify(setID uint64, authorities []NewAuthority[ID]) error {
+func verify[H HashI, N constraints.Unsigned, S comparable, ID constraints.Ordered, Header HeaderI[H, N]](j Justification[H, N, S, ID, Header], setID uint64, weights []finalityGrandpa.IDWeight[ID]) error {
+	// TODO Get reviewer feedback on this. In substrate they pass in data then convert to IDWeight, however for
+	// us we do no ever use this input type, so I think better to just directly take IDWeights as a param.
+	// Can revert if people disagree
+	//var weights []finalityGrandpa.IDWeight[ID]
+	//for _, authority := range authorities {
+	//	weight := finalityGrandpa.IDWeight[ID]{
+	//		ID: authority.Key,
+	//	}
+	//	weights = append(weights, weight)
+	//}
+	voters := finalityGrandpa.NewVoterSet[ID](weights)
+	if voters != nil {
+		err := verifyWithVoterSet(j, setID, *voters)
+		return err
+	}
+	return fmt.Errorf("%w", errInvalidAuthoritiesSet)
+}
+
+// Validate the commit and the votes' ancestry proofs.
+func verifyWithVoterSet[H HashI, N constraints.Unsigned, S comparable, ID constraints.Ordered, Header HeaderI[H, N]](j Justification[H, N, S, ID, Header], setID uint64, voters finalityGrandpa.VoterSet[ID]) error {
+	ancestryChain := newAncestryChain[H, N](j.VotesAncestries)
+	commitValidationResult, err := finalityGrandpa.ValidateCommit[H, N, S, ID](j.Commit, voters, ancestryChain)
+	if err != nil || !commitValidationResult.Valid() {
+		return fmt.Errorf("%w: invalid commit in grandpa justification", errBadJustification)
+	}
```

Comment:
```suggestion
	if err != nil {
		return fmt.Errorf("%w: invalid commit in grandpa justification", errBadJustification)
	}
	if !commitValidationResult.Valid() {
		return fmt.Errorf("%w: invalid commit in grandpa justification", errBadJustification)
	}
```

Reply:
nit: I find it easier to read.  I had to check in rust how/when the cases fall through. Not super intuitive.